client/hotrod-client/src/test/java/org/infinispan/client/hotrod/ExternalAddressRoutingTest.java

```diff
@@ -0,0 +1,194 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+
+import org.infinispan.Cache;
+import org.infinispan.client.hotrod.impl.AddressMapper;
+import org.infinispan.client.hotrod.test.HotRodClientTestingUtil;
+import org.infinispan.commons.util.Util;
+import org.infinispan.configuration.cache.CacheMode;
+import org.infinispan.configuration.cache.ConfigurationBuilder;
+import org.infinispan.lifecycle.ComponentStatus;
+import org.infinispan.manager.EmbeddedCacheManager;
+import org.infinispan.server.hotrod.HotRodServer;
+import org.infinispan.util.ControlledConsistentHashFactory;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.Test;
+
+import com.github.terma.javaniotcpproxy.StaticTcpProxyConfig;
+import com.github.terma.javaniotcpproxy.TcpProxy;
+
+@Test(groups = "functional", testName = "client.hotrod.ExternalAddressRoutingTest")
+public class ExternalAddressRoutingTest extends HitsAwareCacheManagersTest {
+
+   private static final int PROXY_TO_SERVER_1_PORT = 9988;
+   private static final int PROXY_TO_SERVER_2_PORT = 9989;
+
+   private static final Map<SocketAddress, SocketAddress> proxyToRealAddresses = new HashMap<>();
+   private static final Map<SocketAddress, SocketAddress> realAddressesToProxy = new HashMap<>();
+
+   private static final String DIST_ONE_CACHE_NAME = "dist-one-cache";
+   private static final String DIST_TWO_CACHE_NAME = "dist-two-cache";
+
+   HotRodServer server1;
+   HotRodServer server2;
+
+   ConfigurationBuilder defaultBuilder;
+   ConfigurationBuilder distOneBuilder;
+   ConfigurationBuilder distTwoBuilder;
+
+   TcpProxy proxyToServer1;
+   TcpProxy proxyToServer2;
+
+   RemoteCacheManager rcm;
+
+   protected ConfigurationBuilder defaultCacheConfigurationBuilder() {
+      return hotRodCacheConfiguration(getDefaultClusteredCacheConfig(CacheMode.DIST_SYNC));
+   }
+
+   @Override
+   protected void createCacheManagers() throws Throwable {
+      defaultBuilder = defaultCacheConfigurationBuilder();
+      distOneBuilder = hotRodCacheConfiguration(getDefaultClusteredCacheConfig(CacheMode.DIST_SYNC));
+      distOneBuilder.clustering().hash().numOwners(1).numSegments(1)
+            .consistentHashFactory(new ControlledConsistentHashFactory(0));
+      distTwoBuilder = hotRodCacheConfiguration(getDefaultClusteredCacheConfig(CacheMode.DIST_SYNC));
+      distTwoBuilder.clustering().hash().numOwners(1).numSegments(1)
+            .consistentHashFactory(new ControlledConsistentHashFactory(1));
+
+      server1 = addHotRodServer();
+      server2 = addHotRodServer();
+
+      blockUntilViewReceived(manager(0).getCache(), 2);
+      blockUntilCacheStatusAchieved(manager(0).getCache(), ComponentStatus.RUNNING, 10000);
+      blockUntilCacheStatusAchieved(manager(1).getCache(), ComponentStatus.RUNNING, 10000);
+
+      StaticTcpProxyConfig proxyToServer1Configuration = new StaticTcpProxyConfig(PROXY_TO_SERVER_1_PORT, "localhost", server1.getPort());
```

Comment:
Wondering if these proxy ports should not be generated randomly making sure the ports are available, otherwise if we run this in a system where ports `9989` or `9988` are taken it would fail.